UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs

```diff
@@ -4,46 +4,141 @@
 #if UNITY_EDITOR
 using UnityEditor;
 #endif
+using System;
+using System.Collections.Generic;
+using System.Linq;
 using UnityEngine;
 using MLAgents.CommunicatorObjects;
+using NUnit.Framework.Constraints;
+using UnityEditor.Experimental.Build.AssetBundle;
 
 namespace MLAgents
 {
     /// Responsible for communication with External using gRPC.
     public class RpcCommunicator : ICommunicator
     {
+        public event QuitCommandHandler QuitCommandReceived;
+        public event ResetCommandHandler ResetCommandReceived;
+        public event RLInputReceivedHandler RLInputReceived;
+
         /// If true, the communication is active.
         bool m_IsOpen;
 
+        /// The default number of agents in the scene
+        private const int k_NumAgents = 32;
+
+        /// Keeps track of which brains have data to send on the current step
+        Dictionary<string, bool> m_HasData =
+            new Dictionary<string, bool>();
+
+        /// Keeps track of which brains queried the batcher on the current step
+        Dictionary<string, bool> m_HasQueried =
+            new Dictionary<string, bool>();
+
+        /// Keeps track of the agents of each brain on the current step
+        Dictionary<string, List<Agent>> m_CurrentAgents =
+            new Dictionary<string, List<Agent>>();
+
+        /// The current UnityRLOutput to be sent when all the brains queried the batcher
+        UnityRLOutputProto m_CurrentUnityRlOutput =
+            new UnityRLOutputProto();
+
+        Dictionary<string, List<Tuple<Agent, AgentAction>>> m_LastActionsReceived =
+            new Dictionary<string, List<Tuple<Agent, AgentAction>>>();
+
+
 # if UNITY_EDITOR || UNITY_STANDALONE_WIN || UNITY_STANDALONE_OSX || UNITY_STANDALONE_LINUX
         /// The Unity to External client.
         UnityToExternalProto.UnityToExternalProtoClient m_Client;
 #endif
         /// The communicator parameters sent at construction
-        CommunicatorParameters m_CommunicatorParameters;
+        CommunicatorInitParameters m_CommunicatorInitParameters;
         
         /// <summary>
         /// Initializes a new instance of the RPCCommunicator class.
         /// </summary>
-        /// <param name="communicatorParameters">Communicator parameters.</param>
-        public RpcCommunicator(CommunicatorParameters communicatorParameters)
+        /// <param name="communicatorInitParameters">Communicator parameters.</param>
+        public RpcCommunicator(CommunicatorInitParameters communicatorInitParameters)
         {
-            m_CommunicatorParameters = communicatorParameters;
+            m_CommunicatorInitParameters = communicatorInitParameters;
         }
+        
+        #region Initialization
+        
         /// <summary>
-        /// Initialize the communicator by sending the first UnityOutput and receiving the
-        /// first UnityInput. The second UnityInput is stored in the unityInput argument.
+        /// Sends the academy parameters through the Communicator.
```

Comment:
nit: seems a little scattered on the terminology. "Sends the academy parameters through" but takes a "CommunicatorInitParameters". And it returns "UnityRLInitParameters" but that's used as "var inputParameters" in the Academy setup.

Reply:
I'll make it more consistent

Reply:
updated.  Let me know if that is better.